src/social/api.ts

```diff
@@ -251,237 +312,287 @@ export const removePost = (
+ *
+ * @param a one post command
+ * @param b another post command
+ *
+ * @returns true if the provided post commands are equal, false otherwise
+ *
+ * @remarks This is a logical comparison from the perspective of the ordering
+ * of the command log, so it is possible that two different post commands
+ * objects are equal.
+ */
+export const arePostCommandIdsEqual = (a: PostCommandId, b: PostCommandId): boolean =>
+    a.timestamp === b.timestamp && a.source === b.source;
+
+/**
+ * Returns the highest seen (Lamport) timestamp from a log
+ *
+ * @param postCommandLog the log
+ *
+ * @returns the highest seen timestamp or 0 if the log is empty
+ */
+export const getHighestSeenTimestampFromLog = (postCommandLog: PostCommandLog): number => {
+    if (postCommandLog.commands.length === 0) {
+        return 0;
     }
-    const timestamp = getLatestPostCommandTimestampFromLog(postCommandLog) + 1;
-    const previousEpoch = getPreviousCommandEpochFromLog(postCommandLog);
-
-    const removedPost: Post = {
-        _id: post._id,
-        text: '',
-        images: [],
-        createdAt: post.createdAt,
-    };
-    const postCommand: PostCommand = {
-        protocolVersion: PostCommandProtocolVersion,
-        post: removedPost,
-        type: 'remove',
-        source,
-        timestamp,
-        parentTimestamp,
-        previousEpoch,
-    };
-    const removedPostCommand =  await addPostCommandToFeed(postCommand, swarmFeedApi);
-    return {
-        ...postCommandLog,
-        commands: [removedPostCommand, ...postCommandLog.commands],
-    };
+    if (postCommandLog.commands[0].epoch != null) {
+        return postCommandLog.commands[0].id.timestamp;
+    }
+    const highestUnsyncedTimestamp = postCommandLog.commands[0].id.timestamp;
+    if (highestUnsyncedTimestamp === postCommandLog.commands.length) {
+        // never synced
+        return highestUnsyncedTimestamp;
+    }
+    for (const command of postCommandLog.commands) {
+        if (command.epoch != null) {
+            if (command.id.timestamp > highestUnsyncedTimestamp) {
+                return command.id.timestamp;
+            } else {
+                break;
+            }
+        }
+    }
+    return highestUnsyncedTimestamp;
 };
 
-const addPostCommandToFeed = async (postCommand: PostCommand, swarmFeedApi: Swarm.FeedApi): Promise<PostCommand> => {
-    const feedTemplate = await swarmFeedApi.downloadFeedTemplate();
-    const updatedCommand = {
-        ...postCommand,
-        epoch: feedTemplate.epoch,
-    };
-    const data = serialize(updatedCommand);
-    await swarmFeedApi.updateWithFeedTemplate(feedTemplate, data);
-    return updatedCommand;
+/**
+ * Returns the epoch of the previous (first) command from the log
+ *
+ * @param postCommandLog the log
+ *
+ * @returns the epoch or undefined if the log is empty
+ */
+export const getPreviousCommandEpochFromLog = (postCommandLog: PostCommandLog): Swarm.Epoch | undefined => {
```

Comment:
[Don't use undefined as a means of denoting validity](https://basarat.gitbooks.io/typescript/docs/javascript/null-undefined.html)
I agree with this.

Reply:
I use this pattern at several places and at some places we use nulls as well, so let's discuss this further in the code conventions discussion #150 .

Also in the `PostCommand` when `epoch` is `undefined` it means that it hasn't been uploaded and this is consistent throughout the API. I am open to suggestions to improve this.